azkaban-common/src/main/java/azkaban/executor/ExecutableRamp.java

```diff
@@ -51,226 +59,106 @@
-
-    public boolean isPaused() {
-      return isPaused;
-    }
-
-    public State setPaused(boolean paused) {
-      this.isPaused = paused;
-      return this;
-    }
-
-    public int getRampStage() {
-      return rampStage;
-    }
-
-    public boolean isRamping() {
-      return (rampStage >= 0);
-    }
-
-    public State setRampStage(int rampStage) {
+      this.numOfTrail = numTrail;
+      this.numOfSuccess = numSuccess;
+      this.numOfFailure = numFailure;
+      this.numOfIgnored = numIgnored;
+      this.isPaused = isPaused;
       this.rampStage = rampStage;
-      return this;
-    }
-
-    public boolean isActive() {
-      return isActive;
-    }
-
-    public State setActive(boolean active) {
-      this.isActive = active;
-      return this;
-    }
-
-    public boolean isSynchronized() {
-      return isSynchronized;
+      this.isActive = isActive;
     }
 
-    private State markChanged() {
-      this.isSynchronized = false;
-      return this;
-    }
-
-    public State markDataSaved() {
-      int aggNumOfTrail = this.numOfTrail + this.cachedNumOfTrail;
-      int aggNumOfSuccess = this.numOfSuccess + this.cachedNumOfSuccess;
-      int aggNumOfFailure = this.numOfFailure + this.cachedNumOfFailure;
-      int aggNumOfIgnored = this.numOfIgnored + this.cachedNumOfIgnored;
-      setNumOfTrail(aggNumOfTrail);
-      setNumOfSuccess(aggNumOfSuccess);
-      setNumOfFailure(aggNumOfFailure);
-      setNumOfIgnored(aggNumOfIgnored);
-      this.cachedNumOfTrail = 0;
-      this.cachedNumOfSuccess = 0;
-      this.cachedNumOfFailure = 0;
-      this.cachedNumOfIgnored = 0;
-      this.isSynchronized = true;
-      return this;
+    public static final State createInstance(long startTime, long endTime, long lastUpdatedTime,
+        int numTrail, int numSuccess, int numFailure, int numIgnored,
+        boolean isPaused, int rampStage, boolean isActive) {
+      return new State(startTime, endTime, lastUpdatedTime,
+          numTrail, numSuccess, numFailure, numIgnored,
+          isPaused, rampStage, isActive);
     }
 
     @Override
     public State refresh(State source) {
-
       this.startTime = source.startTime;
-      this.endTime = (source.endTime > this.endTime) ? source.endTime : this.endTime;
-      this.lastUpdatedTime = (source.lastUpdatedTime > this.lastUpdatedTime) ? source.lastUpdatedTime : this.lastUpdatedTime;
+      this.endTime = source.endTime;
+      this.lastUpdatedTime = source.lastUpdatedTime;
 
       this.numOfTrail = source.numOfTrail;
       this.numOfSuccess = source.numOfSuccess;
       this.numOfFailure = source.numOfFailure;
-      this.numOfIgnored = source.numOfFailure;
+      this.numOfIgnored = source.numOfIgnored;
 
       this.isPaused = source.isPaused;
-      this.rampStage = (source.rampStage > this.rampStage) ? source.rampStage : this.rampStage;
+      if (source.rampStage > this.rampStage) {
+        this.lastRampDownTime = 0;
+      }
+      this.rampStage = source.rampStage;
```

Comment:
As I explained in previous comments, it is a bug which I am fixing. 
There is ramp status control feature : 'When there are more workflow failures detected during the ramp on particular stage, we had better to auto ramp down from Stage N to Stage N-1 to limit the disaster'

The original code has a bug. It will block the action to manually ramp down because when  source.rampStage < this.rampStage, the rampStage didn't get updated.
Right now, after introducing the lastRampDownTime, the problem can be addressed. The rampStage can be updated directly. However, to make sure any immediately ramp up can be automatically ramp down when more error detected, lastRampDownTime need to be reset. If you need more detail information, we can discuss that offline.